### DISCUSSION TITLE
Trying to migrate to latest version, but many things moved or are missing.

### DISCUSSION THREAD

**Discussion by woutware** [SixLabors/Fonts#236] [Q&A]
Class RendererOptions missing, what's the replacement? I guess TextOptions?
Constructor new TextOptions(Font, Vector2) missing (was there for RendererOptions).
Property Font.EmSize is missing, can't find them in Font.FontMetrics.
Class IReadonlyFontCollectionExtensions is entirely missing. I used to call IReadonlyFontCollectionExtensions.CreateFont to create a font. What replaced this? I'm guessing just call new Font(...).
Class GlyphMetric is missing, I'm guessing GlyphBounds.

You can make it easier on users by leaving the classes/methods/properties where they are and marking them [Obsolete] with a description of what the user is to be using instead. And then leave it there for 5 years or so to give us plenty of time to migrate. Not everyone is updating their dependencies multiple times a year.

**Comment by JimBobSquarePants** [comment]
@woutware I've converted this to a discussion since it's not an issue. I'm honestly a little surprised and disappointed that you chose to delete the issue template.

First things first. This is beta software. We don't have a stable API and we explicitly call that out in the documentation. 
>[Fonts is still considered BETA quality and we still reserve the rights to change the API shapes.](https://docs.sixlabors.com/articles/fonts/index.html?tabs=tabid-1)

We wouldn't and couldn't add `[Obsolete]` to beta software.

Now to answer your question.

- [`RendererOptions`](https://docs-v1.sixlabors.com/api/Fonts/SixLabors.Fonts.RendererOptions.html) => [`TextOptions`](https://docs.sixlabors.com/api/Fonts/SixLabors.Fonts.TextOptions.html). You are looking for the `Origin` property. Since the only required property is `Font` it was deemed odd and unnecessary to preserve the additional constructor. We renamed this property when we streamlined the interaction between Fonts and ImageSharp.Drawing removing a duplicate type which required copying properties from the former.
- [`Font.EmSize`](https://docs-v1.sixlabors.com/api/Fonts/SixLabors.Fonts.Font.html#SixLabors_Fonts_Font_EmSize) => [`FontMetrics.UnitsPerEm`](https://docs.sixlabors.com/api/Fonts/SixLabors.Fonts.FontMetrics.html#SixLabors_Fonts_FontMetrics_UnitsPerEm) This name is more in keeping with the [specification](https://docs.microsoft.com/en-us/typography/opentype/spec/ttch01#funits-and-the-grid). 
- [`IReadonlyFontCollectionExtensions.CreateFont`](https://docs-v1.sixlabors.com/api/Fonts/SixLabors.Fonts.IReadonlyFontCollectionExtensions.html#methods) => [`FontFamily.CreateFont`](https://docs.sixlabors.com/api/Fonts/SixLabors.Fonts.FontFamily.html#methods) Font families are created based upon the given culture from [`FontCollection`](https://docs.sixlabors.com/api/Fonts/SixLabors.Fonts.FontCollection.html#SixLabors_Fonts_FontCollection_Get_System_String_System_Globalization_CultureInfo_) hence the lack of overloads accepting one. The refactor was designed to better group functionality to individual functional components.
- [`GlyphMetric`](https://docs-v1.sixlabors.com/api/Fonts/SixLabors.Fonts.GlyphMetric.html) => [`GlyphBounds`](https://docs.sixlabors.com/api/Fonts/SixLabors.Fonts.GlyphBounds.html) This name better describes the type.

Every single change has been discussed and made in the public domain with the relevant PRs linked in the release notes.

**Comment by woutware** [comment]
Thank you for the quick reply and clarifying the changes.

I wanted to submit an issue rather than create a discussion as it was an issue for me. My issue just didn't fit the template, that's all.

Anyways, I would argue that especially in Beta the [Obsolete] tag is useful, since that's when API changes will happen mostly. Doesn't that make sense? For each of the changes above I would have taken minutes in making all the adjustments rather than checking out diffs on github for half a day. I don't see much point in adopting the [Obsolete] tag when the API has settled down, what's the rationale of not using it in beta? Also this package has been in beta for years, without any date on a release date.

**Reply by JimBobSquarePants** [reply]
>Doesn't that make sense?

That would require the change to be small enough to allow the original method or property to remain. The stability expectation of beta software is subject to opinion but holds no guarantees. For us, beta means externally testable but not feature complete - That's why we  use the disclaimer - And as such we might not be able to maintain a path of continuation in order to do so.

I'd generally use [Obsolete] sparingly, when [I absolutely have to](https://docs-v1.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.DependencyInjection.ImageSharpBuilderExtensions.html#SixLabors_ImageSharp_Web_DependencyInjection_ImageSharpBuilderExtensions_SetMemoryAllocator_SixLabors_ImageSharp_Web_DependencyInjection_IImageSharpBuilder_System_Func_System_IServiceProvider_MemoryAllocator__). Once a major version becomes stable we'll try very, very hard not to break things.

Fonts needed a LOT of work to bring it up to what I would consider usable quality. Unfortunately that meant breaking a lot of things. I would confidently say now though that the dust has settled and we're very close to approaching RC candidate quality though I'd like to get some of [these issues](https://github.com/SixLabors/Fonts/issues?q=is%3Aissue+is%3Aopen+label%3Ashaping) closed off first.

> Also this package has been in beta for years, without any date on a release date.

There's so, so few [contributors](https://github.com/SixLabors/Fonts/graphs/contributors) and writing a font loading and shaping API is really hard.

**Reply by woutware** [reply]
I don't mind the beta label, I don't mind changes either.

But it's next to zero effort to leave the Font.EmSize property there with an [Obsolete("This moved to bla bla.")] tag. Same for the CreateFont call etc. It's just a matter of scalability. As your number of users grows and you just have a handful of contributors, you don't want hundreds of users searching for where the cheese moved, when you have such a perfect tool given by .NET to handle this.

Also why use [Obsolete] sparingly? As a user I'd prefer to see it as much as possible. I don't mind if I bump into them dozens of times, the compiler will tell me what to do and I don't need to browse github for hours.

Anyways, I suggest you reconsider the restrictions on usage of [Obsolete] as it's very very useful, and it especially scales well towards many users.

**Reply by antonfirsov** [reply]
@woutware managing obsolation between beta versions is not realistic with our resources.

Some context:

SixLabors is still not monetizing at any meaningful scale, despite all the efforts invested by James. This means that everything we do we do in our limited free time, that makes things much harder. An example: one of the anticipated changes is the removal of `RenderOptions` which I initiated in #212 after a 30 minutes review of the Fonts code somewhere around the end of my day. I was super happy James was able to react to that and we managed to clean up the API in a series of similar minimal-effort sessions. These kind of cosmetics are on the bottom of our backlog, since the the top priority is to deliver features at high quality. It would practically kill our API improvement efforts if we would also have to spend time to find proper obsoletion strategies in a complicated case where we not only rename or move properties, but also remove types. We will never get out of beta state with such expectations, and the *real problem* is that Drawing + Fonts are still in beta.

Being more understanding with such changes between releases of a beta version is a sort-of community help, and we are in a very high need of community help on our journey maintaining the SixLabors libraries.

An alternative to community-centric approach is to own a support license. It definitely won't bribe us to change our mind on `[Obsolete]`, but it may put you to a better position when it's about getting help in cases breaking changes occur. (But I shut up on this front, since this is managed entirely by James :)

----

Also giving heads-up about further breaking changes that will land in ImageSharp.Drawing before it reaches stable 1.0: https://github.com/SixLabors/ImageSharp.Drawing/issues/195

**Reply by Audionysos** [reply]
I don't get this. Obsoletion strategies? He didn't mention any complex cases. What kind of "strategy" do you need to write `[Obsolete("Moved to FontMetrics.UnitsPerEm")]` or `[Obsolete("Use FontFamily.CreateFont")]`? How much time exactly did you saved on this, considering the fact that you actually wrote this anyway in this discussion, with also the reasoning behind the changes?
I don't think people are interested about your reasoning, as long as we can quickly find the alternatives. 

"That would require the change to be small enough to allow the original method or property to remain."

Why? Do you think you should keep it working the same way? Just leave the signature, clear the contents and put `throw new ObsoleteException()`. It's not the problem that library under development introduces breaking changes, because users expect them. The problem is that users don't know how to address those changes.

"...in a complicated case where we not only rename or move properties, but also remove types."
What is the problem with removing type? Just don't remove it, mark it obsolete. Eventually put the file in "trash" folder.

I think also we are not expecting for you to document every little single change you made, but something like creating a font in a fonts library is just fundamental, so if you changed this, you may be sure that 100% of your users have broken applications after the update... 

"...It definitely won't bribe us to change our mind on [Obsolete], but it may put you to a better position when it's about getting help in cases breaking changes occur."
The point is that if you would document a change, no one would need getting help on it. Your users, paid or not, wouldn't lose the time, we could test other areas of the lib. You also would not lose time on giving the help and could spend it on the actual development... Unless your strategy is to make your product hard to use and getting money by providing the support... :P

**Reply by JimBobSquarePants** [reply]
It was a beta.

**Comment by Audionysos** [comment]
I actually wanted to ask a question.

I'm interested in acquiring glyph's data (shapes).  Previously (version 0.0...don't remember) I was doing it like this:
```C#
using F = SixLabors.Fonts;
using X = com.audionysos.text.render;

namespace WpfDNet {
	public class SLGlyphsProvider : GlyphsProvider {

		public override X.Glyph get(char ch, ITextFormat f) {
			if (ch == ' ') return missingGlyph;
			var ff = F.SystemFonts.Find(f.font.name);
			var font = ff.CreateFont((float)f.size);
			var gl = font.GetGlyph(ch);
			var gi = gl.Instance;
			//gi.
			var min = gi.Height - font.EmSize;
			var scale = (float)f.size / 11; var x = 0; var y = 0;
			var sf = (font.Ascender + font.Descender) / gi.ScaleFactor * scale;
			//var sf = (font.Ascender) / gi.ScaleFactor * scale;
			var s = new System.Numerics.Vector2(sf, -sf);
			var off = new System.Numerics.Vector2(0 + x, (gi.Height - min) * s.X + y);

			var cp = gi.ControlPoints[0] * s + off;
			var fp = cp;

			var color = 0u;
			var width = 0; var height = 0;

			var phs = new List<X.Path>();
			var ph = new X.Path();
			phs.Add(ph);

			ph.Add(cp.X, cp.Y);
			var nep = gi.EndPoints[0]; var epi = 0;
			var m = false;
			for (int i = 1; i < gi.ControlPoints.Length; i++) {
				cp = gi.ControlPoints[i] * s + off;
				if (m) {
					ph = new X.Path();
					phs.Add(ph);
					ph.Add(cp.X, cp.Y);
					m = false;
					fp = cp;
					continue;
				}
				ph.Add(cp.X, cp.Y);
				if (nep == i) {
					//g.lineTo(fp.X, fp.Y);
					epi++;
					if (epi < gi.EndPoints.Length)
						nep = gi.EndPoints[epi];
					m = true;
				}
			}

			var r = new X.Glyph(width, height, phs);
			return r;
		}

	}
}
```

Now I can't find  the `ControlPoints`/`Endpoints`. I started digging into source code but eventually gave up and implemented the `IGlyphRenderer`: 

```C#
using F = SixLabors.Fonts;
using SixLabors.Fonts;
using System.Numerics;
using X = com.audionysos.text.render;
using System.Collections.Generic;
using System;

namespace WpfDNet {
	public class bla : F.IGlyphRenderer {
		public List<X.Path> phs = new ();
		private X.Path ph = null;
		private Vector2 cp;

		public TextDecorations EnabledDecorations() {
			return TextDecorations.None;
		}

		public void SetDecoration(TextDecorations textDecorations, Vector2 start, Vector2 end, float thickness)
			{}

		public void BeginText(in FontRectangle bounds) {}

		public bool BeginGlyph(in FontRectangle bounds, in GlyphRendererParameters parameters)
			=> true;

		public void BeginFigure() {
			//ph = new X.Path ();
			cp = new Vector2();
		}

		private float s = 1;
		public void LineTo(Vector2 p) {
			cp = p * s;
			ph.Add(cp.X, cp.Y);
		}

		public void MoveTo(Vector2 point) {
			if(ph != null) phs.Add(ph);
			ph = new X.Path();
			cp = point * s;
			ph.Add(cp.X, cp.Y);
		}

		public void QuadraticBezierTo(Vector2 c, Vector2 e) {
			var g = 10f;
			for (int i = 1; i <= g; i++) {
				var t = i / g;
				var m1 = mid(cp, c, t);
				var m2 = mid(c, e, t);
				cp = mid(m1, m2, t);
				cp *= s;
				ph.Add(cp.X, cp.Y);
			}
		}

		private Vector2 mid(Vector2 s, Vector2 e, float t)
			=> s + (e - s) * t;

		public void CubicBezierTo(Vector2 c, Vector2 c2, Vector2 e) {
			var g = 10f;
			for (int i = 0; i <= g; i++) {
				var t = i / g;
				var m1 = mid(cp, c, t);
				var m2 = mid(c, c2, t);
				var m3 = mid(c2, e, t);
				var m21 = mid(m1 , m2, t);
				var m22 = mid(m2 , m3, t);
				cp = mid(m21, m22, t);
				cp *= s;
				ph.Add(cp.X, cp.Y);
			}
		}

		public void EndFigure() {
			//var sp = ph[0];
			//ph.Add(sp);
			phs.Add(ph);
		}

		public void EndGlyph() {}

		public void EndText() {}

	}
}

using com.audionysos.text.edit;
using com.audionysos.text.render;
using SixLabors.Fonts;
using System.Collections.Generic;
using F = SixLabors.Fonts;
using X = com.audionysos.text.render;

namespace WpfDNet {
	public class SLGlyphsProvider : GlyphsProvider {

		public override X.Glyph get(char ch, ITextFormat f) {
			if (ch == ' ') return missingGlyph;
			var font = F.SystemFonts.CreateFont(f.font.name, (float)f.size);
			font.TryGetGlyphs(new F.Unicode.CodePoint(ch), out var glyphs);
			var g = glyphs[0];
			var b = new bla();
			TextRenderer.RenderTextTo(b,  ch.ToString(), new TextOptions(font) { 
				
			});
			var rc = g.BoundingBox(GlyphLayoutMode.Horizontal
				,new System.Numerics.Vector2(), 128);
			var width = rc.Width; var height = 0;
			var r = new X.Glyph(width, height, b.phs);
			return r;
		}
	}

}

```

So now output of my glyph provider shouldn't be different. However I have problem with the holes:
  
![sixlabors_fonts_problem](https://github.com/SixLabors/Fonts/assets/26059765/9cc8dbe3-ff5e-4866-a087-9c02a8d46561)

I have no idea what is happening. I don't know If I implemented the `IGlyphRenderer` wrong or is it something in the update of SixLabors.Drawing library. Especially the colon is strange. Why one dot is fille and the other is not? From what I remember a shape was a hole if it was drawn inside another shape, right? 
I didn't change much except `drawLines()` => `drawLine()` `IBrush` => `Brush` etc.
I even have a comment left in my shapes converter about the holes in glyphs "figures with the same brush are joined so the ImageSharp will create holes (rendering glyphs)"... Those paths from `IGlyphRenderer` are passed through my graphics abstraction layer but eventually they are passed back to SixLabors:
```C#
		#region Image sharp rendering
		/// <summary>ImageSharp ready figures</summary>
		private List<SharpFigure> sFigures = new List<SharpFigure>();
		private void covertToSharpFigures() {
			sFigures.Clear();
			var joined = new List<Figure>(); //figures with the same brush are joined so the ImageSharp will create holes (rendering glyphs)
			Figure last = null;
			for (int i = 0; i < tFigures.Count; i++) {
				var tf = tFigures[i];
				if (last == null || tf.sameStyle(last)) {
					joined.Add(tf);
					last = tf;
					continue;
				}
				var sf = new SharpFigure(joined, last.fill, last.stroke);
				sFigures.Add(sf);
				joined.Clear(); last = null;
			}
			if (joined.Count == 0) return;
			sFigures.Add(new SharpFigure(joined, last.fill, last.stroke));
		}

		public void render(Image<Bgra32> img) {
			img.Mutate(x => {
				for (int i = 0; i < sFigures.Count; i++) {
					var f = sFigures[i];
					if (f.brush != null)
						x.Fill(f.brush, f.path);
					if (f.pen == null) continue;
					for (int j = 0; j < f.points.Length; j++) {
						//x.DrawLines(f.pen, f.points[j]);
						x.DrawLine(f.pen, f.points[j]);
					}
				}
			});
		}
		#endregion
``` 
In case you would want to look, you can find above code here:
[GlyphsProvider](https://github.com/Audionysos/DFM/blob/master/WpfDNet/SixLabors/Graphics/SLGlyphsProvider.cs)
[IGlyphRenderer](https://github.com/Audionysos/DFM/blob/master/WpfDNet/SixLabors/Graphics/bla.cs)
[convert/render](https://github.com/Audionysos/DFM/blob/master/WpfDNet/SixLabors/Graphics/SharpGraphics.cs)

I don't expect you to dig in my code or solve my issues but I don't know were to go from here. This is my spare time project I didn't touched for two years or so and now when I finally got inspired to push it forward I've already spend two days trying to make it work as it was already... :| I understand this was beta version but I hope you would understand me too and maybe give me some pointers.
Also, are you planning to improve the docs? There are only few examples. Some API descriptions are very vague. For example does the glyph figure always closed? Do I need to close the manually? It's tedious to find this empirically...

**Reply by Audionysos** [reply]
Ok, So I dropped my Implementation of `IGlyphRenderer` and used `TextBuilder` to produce `IPath`s from where I get the points:
```C#
		public override X.Glyph get(char ch, ITextFormat f) {
			if (ch == ' ') return missingGlyph;
			var font = F.SystemFonts.CreateFont(f.font.name, (float)f.size);

			var phs = SixLabors.ImageSharp.Drawing.TextBuilder
				.GenerateGlyphs(ch.ToString(), new TextOptions(font)
			);
			var paths = new List<Path>();
			foreach (var ph in phs) {
				var fp = ph.Flatten();
				foreach (var sp in fp) {
					var points = sp.Points.ToArray();
					var path = new Path();
					foreach (var p in points)
						path.Add(p.X, p.Y);
					paths.Add(path);
				}
			}

			font.TryGetGlyphs(new F.Unicode.CodePoint(ch), out var glyphs);
			var g = glyphs[0];
			var rc = g.BoundingBox(GlyphLayoutMode.Horizontal
				, new System.Numerics.Vector2(), 128);
			var width = rc.Width; var height = 0;

			var r = new X.Glyph(width, height, paths);
			return r;
		}
```
I didn't change anything else and here is the result:
![sixlabors_fonts_problem2](https://github.com/SixLabors/Fonts/assets/26059765/6c10e9d8-c7b3-43f2-b7b0-0a261e480ac6)

So, I guess, conclusion is that there was something with my `IGlyphRenderer` output that the `ComplexPolygon` didn't like.
